### PR TITLE
My solution for nearest coffee shops finder

### PR DESCRIPTION
**Pull request by edustef**
## What was done
- Input validation with zod
- Env validation on startup
- Sorting done using euclidean distance (I skipped square root because we only need it to compare, we don't need the actual distances)
- Tests for sorting logic, edge cases and error handling

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -1,12 +1,36 @@
...
+  return shops.toSorted(byDistance(position)).slice(0, MAX_RESULTS);
+}
+
+/**
+ * @param {Position} position
+ * @returns {(a: Position, b: Position) => number}
+ */
+function byDistance(position) {
+  return (a, b) => squaredDistance(position, a) - squaredDistance(position, b);
+}
```
Comment (adrianbunea) [inline src/app.js:24]: I appreciate trying to make the code more idiomatic by naming this function like this. The code above does read much closer to human language.

File: src/app.js [hunk 2]
```diff
@@ -1,12 +1,36 @@
...
-  return [];
+
+/**
+ * @param {Position} position
+ * @returns {Promise<Position[]>}
+ */
+export async function getNearestShops(position) {
+  const shops = await fetchShops()
+
+  return shops.toSorted(byDistance(position)).slice(0, MAX_RESULTS);
```
Comment (adrianbunea) [inline src/app.js:15]: I see you're keeping up with the latest APIs. Now that `toSorted` is introduced, would you ever use the old sort ever again?

File: src/env.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+if (!process.env.API_TOKEN) {
+		throw new Error('API_TOKEN is not defined in environment variables');
```
Comment (adrianbunea) [inline src/env.js]: And how is the user of this script supposed to populate the API_TOKEN environment variable? Docs don't mention anything, and it puts too much work into the hands of the user.

File: src/utils.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  const url = 'https://api-challenge.agilefreaks.com/v1/coffee_shops';
+  const urlWithToken = `${url}?token=${(apiToken)}`;
+
+  const response = await fetch(urlWithToken, {
+    headers: { 'Accept': 'application/json'  },
+  });
+
+  if (!response.ok) {
+    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
+  }
```
Comment (adrianbunea) [inline src/utils.js:40]: Feels a bit abrupt to call it quits just because the request failed. The script could be much more resilient. What can you do to improve it?

File: test/app.test.js [hunk 5]
```diff
@@ -1,10 +1,72 @@
...
-describe('App', () => {
-  it('should return an array when the input is valid', () => {
-    expect(Array.isArray(getNearestShops({
-      lat: 0,
-      lng: 0,
-    }))).toBe(true);
+jest.mock('../src/utils', () => ({
+  ...jest.requireActual('../src/utils'),
+  fetchShops: jest.fn(),
+}));
```
Comment (adrianbunea) [inline test/app.test.js]: You are mocking your own implementation. If there was a bug in fetchShops these tests would not catch it. And there is quite a bit of logic inside fetchShops that does more than just return a list of shops.

What would be a better way to test your logic while mocking data?

**Review by adrianbunea — COMMENTED** [review]
Hey Eduard. Thanks for your time and effort when working on this. I left a review with some questions, feel free to make code changes or just answer in the replies.

File: src/app.js (hunk 1 above)
Comment (edustef) [reply src/app.js:24]: Thank you! Yeah naming things clearly is important. It means less comments and easier to scan and read.

File: src/app.js (hunk 2 above)
Comment (edustef) [reply src/app.js:15]: I would use `toSorted` most of the times by default and prefer to use non-mutating functions in most use cases. The only reason I would consider otherwise is where performance and memory is very important. For a typical use case like this, `toSorted` is the better choice.

File: src/utils.js (hunk 4 above)
Comment (edustef) [reply src/utils.js:40]: You're right. Here we could write a simple retry function or use a library that does that for us. But a simple fetch wrapper function where you pass a `maxRetries` in this case should be good enough. Is this what you were thinking of?

File: test/app.test.js (hunk 5 above)
Comment (edustef) [reply test/app.test.js]: Probably I would restructure the `getNeasterShops` and extract the `fetchShops` outisde and pass data like so `getNearestShops(data, position)`. Then we can test `fetchShops` separately with something like `msw` or `jest.spyOn(global, 'fetch')`

File: src/env.js (hunk 3 above)
Comment (edustef) [reply src/env.js]: I updated it so that the token is fetched programatically now.

File: src/utils.js (hunk 4 above)
Comment (adrianbunea) [reply src/utils.js:40]: Yeah, retries and recovery (for example fetching token in case of a 401)
